### PR TITLE
feat: Adds support for custom suppliers in AWS and Identity Pool credentials

### PR DESCRIPTION
**Pull request by aeitzman**
Features from #1494 and #1484, also adds documentation.

### CODE REVIEW

File: docs/user-guide.rst [hunk 1]
```diff
@@ -486,6 +486,91 @@ they do not meet your specific requirements.
...
+    https://cloud.google.com/iam/docs/best-practices-for-using-workload-identity-federation#provider-audience
+.. _credential configuration file with the gcloud CLI:
+    https://cloud.google.com/sdk/gcloud/reference/iam/workload-identity-pools/create-cred-config
+
+Accessing resources using a custom supplier with AWS
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This library also allows for a custom implementation of :class:`google.auth.aws.AwsSecurityCredentialsSupplier`
+to be specificed when creating a :class:`google.auth.aws.Credential`. The supplier must
+return valid AWS security credentials, which will then be exchanged for a
```
Comment (clundin25) [inline docs/user-guide.rst:535]: I remember that the custom suppliers _could_ throw an exception.

I think how to handle failures should be documented.

File: docs/user-guide.rst [hunk 2]
```diff
@@ -486,6 +486,91 @@ they do not meet your specific requirements.
...
+This library also allows for a custom implementation of :class:`google.auth.identity_pool.SubjectTokenSupplier`
+to be specificed when creating a :class:`google.auth.identity_pool.Credential`. The supplier must
+return a valid OIDC or SAML2.0 subject token, which will then be exchanged for a
+Google Cloud access token.::
+
+    from google.auth import identity_pool
+
+    class CustomSubjectTokenSupplier(identity_pool.SubjectTokenSupplier):
+
+        def get_subject_token(self, context, request):
```
Comment (clundin25) [inline docs/user-guide.rst:502]: I think it would be helpful to define what the context and requests params are and how to use them.

**Review by clundin25 — APPROVED** [review]

File: google/auth/aws.py [hunk 3]
```diff
@@ -722,30 +598,231 @@ def _get_metadata_role_name(self, request, imdsv2_session_token):
...
-            environment_vars.AWS_DEFAULT_REGION
-        ):
-            return True
 
-        # AWS security credentials can be retrieved from the AWS_ACCESS_KEY_ID
-        # and AWS_SECRET_ACCESS_KEY environment variables. The metadata server
-        # should be used if either of these are not available.
-        if not os.environ.get(environment_vars.AWS_ACCESS_KEY_ID) or not os.environ.get(
-            environment_vars.AWS_SECRET_ACCESS_KEY
+class Credentials(external_account.Credentials):
```
Comment (lsirac) [inline google/auth/aws.py:602]: Shouldn't this be at the top of the file (where it was)?

File: google/auth/aws.py [hunk 4]
```diff
@@ -722,30 +598,231 @@ def _get_metadata_role_name(self, request, imdsv2_session_token):
...
+        )
 
     def _create_default_metrics_options(self):
         metrics_options = super(Credentials, self)._create_default_metrics_options()
-        metrics_options["source"] = "aws"
+        if self._has_custom_supplier():
+            metrics_options["source"] = "programmatic"
+        else:
+            metrics_options["source"] = "aws"
         return metrics_options
```
Comment (lsirac) [inline google/auth/aws.py:810]:
```suggestion
        metrics_options["source"] = "aws"
        if self._has_custom_supplier():
            metrics_options["source"] = "programmatic"
        return metrics_options
```

File: google/auth/identity_pool.py [hunk 5]
```diff
@@ -26,48 +26,172 @@
...
+from typing import NamedTuple
 
 from google.auth import _helpers
 from google.auth import exceptions
 from google.auth import external_account
 
 
+class SubjectTokenSupplier(metaclass=abc.ABCMeta):
+    """Base class for subject token suppliers. This can be implemented with custom logic to retrieve
+    a subject token to exchange for a Google Cloud access token. The identity pool credential does
```
Comment (lsirac) [inline google/auth/identity_pool.py:56]: when using Workload or Workforce Identity Federation

**Review by lsirac — CHANGES_REQUESTED** [review]

File: google/auth/aws.py (hunk 3 above)
Comment (aeitzman) [reply google/auth/aws.py:602]: WDYM? The existing file had the utility class (signer) above the credentials class, I just kept with that pattern

**Review by lsirac — APPROVED** [review]

**Review by BigTailWolf — COMMENTED** [review]
LGTM

**Pull request merged**